### DISCUSSION TITLE
Geoid separation wrong sign convention in GGA parser

### DISCUSSION THREAD

**Discussion by twpayne** [XCSoar/XCSoar#1435] [General]
XCSoar versions having the problem
-------------------------------------------------

All for at least the last 14 years, including the current `master` branch (commit b1244d8dc61e808a3fab3f9d14668aa777e91835 at the time of writing).

XCSoar versions not having the problem
-------------------------------------------------

n/a

System information
------------------

n/a

Steps to reproduce the behavior
-------------------------------

These lines:

https://github.com/XCSoar/XCSoar/blob/b1244d8dc61e808a3fab3f9d14668aa777e91835/src/Device/Parser.cpp#L616-L628

state that FLARM uses MSL altitudes and subtract the geoid separation as computed from EGM96.

This is not correct. FLARM always uses elevation above the WGS84 ellipsoid. The subtraction should be removed.

Expected behavior
-----------------

The altitude is correct.

Actual behavior
---------------

The altitude is incorrect by the value of the geoid separation.

Do you have any idea what may have caused this?
-----------------------------------------------

n/a

Do you have an idea how to solve the issue?
-------------------------------------------

Remove the `else` branch of this `if` statement completely.

**Comment by kobedegeest** [comment]
According to the flarm data protocol the --GGA and PGRMZ sentences are used and both of these use MSL altitude. 
So why do you say it sends out the WGS84 altitude? 
(IGC files however do indeed contain the WGS84 ellipsoid elevation but i do not think this is what the flarm sends out) 

And either way this is inside a function that parses the --GGA sentence which should use MSL. So if flarm were to abuse this sentence (which i doubt) it should be fixed elsewhere than just removing the else statement.

**Comment by lyusupov** [comment]
@kobedegeest 

**it is obviously a [PEBKAC](https://en.wiktionary.org/wiki/PEBKAC) type of issue invented by the submitter** -

<img width="655" alt="image" src="https://github.com/XCSoar/XCSoar/assets/5849637/465257d2-1a8e-423c-8cd4-c35d3eed557f">

**Comment by twpayne** [comment]
If your response to bug reports is ad hominem attacks on the bug reporter ... good luck with your project.

**Comment by DanD222** [comment]
> If your response to bug reports is ad hominem attacks on the bug reporter ... good luck with your project.

The last comment was not from an XCSoar developer (I am not one either), so please don't be offended.

**Comment by lordfolken** [comment]
@twpayne out of interest, what made you think the altitude indicated in XCSoar to be wrong?

**Comment by lyusupov** [comment]
@lordfolken 

the statement 

> ... FLARM always uses elevation above the WGS84 ellipsoid ... 

has actually some sense for FLARM **radio**  data emitted by its built-in sub-1GHz ISM band transmitter.

But it has nothing to do with serial data protocol (**FTD-012**)  that FLARM uses for communication with third-party hardware & software, such as XCSoar.

**Comment by lordfolken** [comment]
@lyusupov the [guy](https://github.com/twpayne) has written a lot of gis and geospatial related tools. I suspect he is quite knowledgeable on the topic. ;-)
He went bug hunting for a reason, and while this particular location probably isn't the cause, i"d like to know what made him hunt for this.

**Comment by lyusupov** [comment]
@lordfolken 

Responsible bug reporters are typically very careful before they can make such statements

`All for at least the last 14 years, including the current master branch (commit https://github.com/XCSoar/XCSoar/commit/b1244d8dc61e808a3fab3f9d14668aa777e91835 at the time of writing).`

**With no arguments at all!**

**Comment by twpayne** [comment]
Note that all FLARM radio communication uses ellipsoid height, i.e. my goal is ensure that all instruments are reporting the correct ellipsoid height.

I raised this issue when debugging an issue in a flight instrument that uses XCSoar as a reference implementation. In this particular case, the instrument was reporting an altitude roughly 100m different to other instruments at my testing location. At my testing location the geoid separation is 47m and I suspect that this was being subtracted when it should be added, which neatly accounts for the 100m difference.

From further reading on GGA NMEA sentences, I understand that the GGA sentence includes both geoid (AMSL) altitude and geoid separation. When both geoid altitude and geoid separation are positive, I believe that things look something like this (excuse the crude ASCII art):

```
        -oo-+ <----- aircraft
          ^
          :
          : <------- altitude AMSL (i.e. above geoid, say 400m)
          :
          v
----____----____---- geoid (AMSL)
          ^
          : <------- geoid separation (say 47m)
          v
______--------______ elipsoid
```

So therefore, the aircraft's height above the ellipsoid is 400+47 = 447m.

My understanding of the current XC Soar code is that it subtracts the geoid separation, i.e. reports an ellipsoid altitude of 400-47 = 353m.

Note that I do not have access to the official NMEA specification for GGA sentences. My understanding of the altitudes is based on what I've found online.

**Comment by kobedegeest** [comment]
This is a completely different issue than what you described in your original post. There you said no correction is needed. Now you claim the correction should be there but with different sign. You first claimed altitude is wrong by geod seperation now it is twice the geod seperation. 

This would be the case accross all devices that use GGA sentence, not a flarm specific problem. 

Pictures of your test would also help convince the issue is there. Based on what i find subtraction is the correct operation. Also if you know the seperation in your location you can monitor the nmea output of the flarm in xcsoar and see the value that is send for the seperation (+ or minus value) as well as other devices. 

> Geoidal Separation in Meters. Geoidal separation is the
> difference between the WGS-84 earth ellipsoid and mean
> sea-level.

That is what i find, i take that to be negative is the ellipsoid is below MSL. The GGA sentence is a commenly used convention not some trade secret. (one source https://resources.winsystems.com/software/nmea.pdf)

Are you sure you used gps altitude, flarm also send pressure data and then it could be an issue with wrong qnh settings. 

There might be an issue but it is hard to believe it is in the GGA sentence. 

Best test would be to look at height in xcsoar with flarm as first device and with internal sensors as first device, see if there is a difference. And then look at the NMEA output of the flarm as well.

**Reply by twpayne** [reply]
> This is a completely different issue than what you described in your original post. There you said no correction is needed. Now you claim the correction should be there but with different sign. You first claimed altitude is wrong by geod seperation now it is twice the geod seperation.

Yes, you're correct. I investigated further and determined that the original statement that the ellipsoid height is incorrect remains true, but my proposed fix was incorrect.

> That is what i find, i take that to be negative is the ellipsoid is below MSL. 

I find the opposite, i.e. that the a negative geoid separation means that the ellipsoid is above MSL.

To support my argument, I use the data from GPSes at my location (Switzerland) where I know that MSL is above the ellipsoid.

You can check for your location at https://cddis.nasa.gov/926/egm96/egm96.html. Although geoids vary, they all attempt to model MSL, and so should have the same sign, as long as the geoid separation is not too close to zero.

> Are you sure you used gps altitude, flarm also send pressure data and then it could be an issue with wrong qnh settings.

This is not specific to FLARM. Although FLARM uses ellipsoid heights, and ellipsoid heights have nothing to do with pressure settings, this is about XCSoar's calculation of the ellipsoid height.

**Reply by lordfolken** [reply]
You can enable the NMEA logger in XCSoar, that outputs pure nmea stream to the logs dir. That way its clear what the source sends and how its formatted. This can even be feed back into XCSoar.

**Comment by lordfolken** [comment]
The geoid seperation, is reported as a negative integer. So like in your example -47. 
So what XCSoar does is 400 - -47 which results in 400 + 47 = 447.

**Reply by twpayne** [reply]
> The geoid seperation, is reported as a negative integer. So like in your example -47.

This statement is not correct. The geoid separation varies across the planet. In areas like Germany, Switzerland, and Iceland it is positive because the geoid is above the ellipsoid. In areas like the Indian Ocean it is negative as the geoid is below the ellipsoid.

**Reply by lordfolken** [reply]
OK true

**Comment by kobedegeest** [comment]
https://aprs.gids.nl/nmea/#gga

Found a source that very clearly states the sign convention. This source agrees with what you are saying.

But the title of your report as well as the description is off. It should have been "Geode seperation wrong sign convention in GGA parser"

Changing this will effect all devices that use the GGA sentence. So we should check nmea output of as many devices as possible to see what is actually used.

**Reply by twpayne** [reply]
I've updated the discussion title.

**Comment by kobedegeest** [comment]
I read the code a bit better. 
First of gps.altitude contains AMSL altitude (so the altitude the GGA sentence should give).
Around the code snipet you show first checks for some bugy hardware that sends no altitude value but sends a geod seperation and takes that as altitude. So it goes if some geod_sep number is given and no altidude use geod_seperation as altitude (totaly irrelavant for most harware definitly flarm).

Then you get the else of the first if (the part you pasted in og message). This is thus visited if there is no geode_separation passed along. Then xcsoar assumes that the altidude given is not msl and corrects it with EGM96 geod to get msl. This is done if use_geod is set to true. This can be set to false (this is done for condor). 

So the subtraction you say is wrong is not even done for flarm (according to flarm's data com protocol i can't test flarm). 

And this subtraction is to go from WG84 to amsl.
